### PR TITLE
Issue #4824 - add configuration on RemoteEndpoint for maxOutgoingFrames

### LINKED ISSUE #4824: WebSocket server outgoing message queue memory growth
**Jetty version**
9.4.27.v20200227
**Java version**
OpenJDK 11.0.3
**OS type/version**
Amazon Linux 2
**Description**
We have observed excessive JVM heap growth twice in the last couple days, and a heap dump shows the majority of the growth is due to a backup of outgoing Jetty WebSocket FrameEntries stored in a single org.eclipse.jetty.websocket.common.extensions.ExtensionStack (see heap dump screenshot attached). We are calling the `RemoteEndpoint.sendString(String text, WriteCallback callback)` method to send messages asynchronously to connected websockets. Are there any known scenarios that could cause this to happen, and is there anything we can do to prevent it?
Screen Shot 2020-04-28 at 2 00 40 PM

### PR DESCRIPTION
**Pull request by lachlan-roberts**
Attempt to address issue #4824 at the Jetty WebSocket API level.

If there are more than `maxOutgoingFrames` waiting to send, any additional sends will fail immediately instead of continuing to queue up in the flusher. The callback will be failed immediately but it will not fail the entire ws connection.

This method is called `maxOutgoingFrames` but its not really as the implementation could later on decide to fragment each send from the `WebSocketRemoteEndpoint` into multiple frames.

### CODE REVIEW

**Comment by lachlan-roberts** [comment]
Maybe it would be better if instead we monitor how many frames are queued at a lower level and if this is exceeded we simply just fail the entire connection.

**Comment by joakime** [comment]
I can see the value in this, and I'm good with this implementation.
But I'm wondering if there is a better way to accomplish this.

@gregw @sbordet thoughts?

**Comment by joakime** [comment]
@lachlan-roberts i think this PR needs a refresh / merge from `jetty-9.4.x`

File: jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/WebSocketRemoteEndpoint.java [hunk 1]
```diff
@@ -303,6 +305,19 @@ public void uncheckedSendFrame(WebSocketFrame frame, WriteCallback callback)
...
             batchMode = getBatchMode();
+
+        if (maxNumOutgoingFrames > 0 && frame.isDataFrame())
+        {
+            // Increase the number of outgoing frames, will be decremented when callback is completed.
+            int outgoingFrames = numOutgoingFrames.incrementAndGet();
+            callback = from(callback, numOutgoingFrames::decrementAndGet);
+            if (outgoingFrames > maxNumOutgoingFrames)
+            {
+                callback.writeFailed(new IOException("Exceeded max outgoing frames: " + outgoingFrames + ">" + maxNumOutgoingFrames));
```
Comment (sbordet) [inline jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/WebSocketRemoteEndpoint.java]: Please change the exception from `IOException` to `WritePendingException`.

**Review by sbordet — CHANGES_REQUESTED** [review]
Please add a test that shows that `maxOutgoingFrames` is respected.

File: jetty-websocket/websocket-api/src/main/java/org/eclipse/jetty/websocket/api/RemoteEndpoint.java [hunk 2]
```diff
@@ -141,6 +141,24 @@
      */
     void setBatchMode(BatchMode mode);
 
+    /**
+     * Set the maximum number of frames which allowed to be waiting to be sent at any one time.
+     * The default value is -1, this indicates there is no limit on how many frames can be
+     * queued to be sent by the implementation.
```
Comment (sbordet) [inline jetty-websocket/websocket-api/src/main/java/org/eclipse/jetty/websocket/api/RemoteEndpoint.java]: Please add some text that explains what happens if the limit is exceeded.
What operation will get what exception, etc.

**Review by sbordet — APPROVED** [review]
LGTM - just improve the javadocs.

**Review by sbordet — APPROVED** [review]

**Pull request merged**